### DISCUSSION TITLE
Getting lazy='raise' when filtering on table that is joined twice, contains_eager

### DISCUSSION THREAD

**Discussion by amin-hrp** [sqlalchemy/sqlalchemy#9250] [Usage Questions] [answered]
Hey, I've consistently gotten issues trying to access values on relationships that I've joined and `contains_eager`ed if I filter on them. Is this a bug?

example: 

All objects have `lazy=raise`
```py    
items = (
        session.query(A)
        .options(raiseload("*"))
        .join(A.b.of_type(b_ref))
        .join(
            b_ref.c.of_type(c_ref)
        )
        .filter(
                c_ref.description == 'some_description'
        )
        .join(A.d.of_type(d_ref))
        .join(d_ref.b.of_type(d_to_b_ref))
        .join(d_to_b_ref.c.of_type(d_to_b_to_c))
        .filter(
            d_to_b_ref.description == 'other description'
        )
        .options(
            contains_eager(A.d.of_type(d_ref))
            .contains_eager(
                d_ref.b.of_type(d_to_b_ref)
            )
            .contains_eager(
                d_to_b_ref.c.of_type(d_to_b_to_c_ref)
            )
        )
        .join(A.e.of_type(e_ref))
        .join(e_ref.d.of_type(e_to_d_ref))
        .join(e_to_d_ref.b.of_type(e_to_d_to_b_ref))
        .join(e_to_d_to_b_ref.c.of_type(e_to_d_to_b_to_c_ref))
        .options(
            contains_eager(A.e.of_type(e_ref))
            .contains_eager(e_ref.d.of_type(e_to_d_ref))
            .contains_eager(
                e_to_d_ref.b.of_type(e_to_d_to_b_ref)
            )
            .contains_eager(
                e_to_d_to_b_ref.c.of_type(e_to_d_to_b_to_c_ref)
            )
        )
        .filter(
            e_to_d_to_b_ref.description == 'some other description'
        )
        .all()
    )
```

I've simplified this a little bit, but it should still work similarly to what I actually have. I'm joining to table B and C several times and filtering them on slightly different criteria. Whenever I try to access table B or C through any of the relationships, I get the lazy='raise' issue. Is this expected?

SQLAlchemy version 1.4

**Comment by zzzeek** [comment, marked as answer]
contains_eager() can be difficult to get right so if any of those contains_eagers are not successfully targeting the attribute, it wont be loaded.   it can also be picky and sometimes fail to actually have any way to target some more complex relationship paths (those are bugs, but some are very hard to fix)

you have an immensely complicated query there so it's hard to say where it's going wrong, I'd debug this iteratively using simpler queries to start.  Also i'd consider relying less upon contains_eager() at that level of complexity if possible.

**Reply by zzzeek** [reply]
there's also a simpler reason joined eager and contains_eager can "fail", if a certain object is loaded by more than one relationship path at a time.   The relationship path that matches the object first is the one where subsequent loading rules will be applied.  that's just another path by which "this contains_eager() seems to have been ignored".

**Reply by amin-hrp** [reply]
Is there another way to utilize the relationships along with filtering? I've only seen contains_eager being used for this. I wanted to avoid querying for all items and then filtering within python so I'm not fetching all rows

**Reply by amin-hrp** [reply]
> The relationship path that matches the object first is the one where subsequent loading rules will be applied. that's just another path by which "this contains_eager() seems to have been ignored".

Does that mean if a relationship was filtered out by part of my query but matched the other part, it would have already had `lazy='raise'` applied the first time, so it would be throwing this error?

**Reply by zzzeek** [reply]
sort of, yes, if you have something like A->b.of_type(b_alias_1).also_load(b_alias_1.things) and A->b.of_type(b_alias_2), a B might be considered under one or the other path, but not both.   

I've seen very difficult contains_eager() problems with much less complicated queries than you have up there so I really can't be that confident that you are hitting limitations in contains_eager() vs. there are actual corrections that can be made to your query.   breaking the problem into much smaller pieces and working iteratively is the only way to crack a nut like the one you have up there.

**Reply by amin-hrp** [reply]
gotcha, thank you for clarifying. will try to move off of contains_eager usage

**Reply by amin-hrp** [reply]
I tried switching this over to using just `joinedload` and `selectinload` instead of `contains_eager`. I removed all the filtering as well, still getting the same error, so I think it's the simpler reason you were speaking of.
